zebra-state/src/service/queued_blocks.rs

```diff
@@ -196,4 +196,104 @@ impl QueuedBlocks {
     pub fn utxo(&self, outpoint: &transparent::OutPoint) -> Option<transparent::Utxo> {
         self.known_utxos.get(outpoint).cloned()
     }
+
+    /// Clears known_utxos, by_parent, and by_height, then drains blocks.
+    /// Returns all key-value pairs of blocks as an iterator
+    pub fn drain(&mut self) -> Drain<'_, block::Hash, QueuedNonFinalized> {
+        self.known_utxos.clear();
+        self.known_utxos.shrink_to_fit();
+        self.by_parent.clear();
+        self.by_parent.shrink_to_fit();
+        self.by_height.clear();
+
+        self.blocks.drain()
+    }
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct SentHashes {
+    /// A list of previously sent block batches, each batch is in increasing height order.
+    /// We use this list to efficiently prune outdated hashes that are at or below the finalized tip.
+    bufs: Vec<VecDeque<(block::Hash, block::Height)>>,
+
+    /// The list of blocks sent in the current batch, in increasing height order.
+    curr_buf: VecDeque<(block::Hash, block::Height)>,
+
+    /// Stores a set of hashes that have been sent to the block write task but
+    /// may not be in the finalized state yet.
+    sent: HashMap<block::Hash, Vec<transparent::OutPoint>>,
+
+    /// Known UTXOs.
+    known_utxos: HashMap<transparent::OutPoint, transparent::Utxo>,
+}
+
+impl SentHashes {
+    /// Inserts the block's utxos in `known_utxos` by outpoints, it's outpoints into `sent` by the block hash,
+    /// and pushes (block.hash, block.height) onto the current buf to be used for pruning outdated hashes.
+    ///
+    /// Assumes that hashes are added in the order of their height between `finish_batch` calls
+    /// for efficient pruning.
+    pub fn add(&mut self, block: &PreparedBlock) {
+        // Track known UTXOs in sent blocks.
+        let outpoints = block
+            .new_outputs
+            .iter()
+            .map(|(outpoint, ordered_utxo)| {
+                self.known_utxos
+                    .insert(*outpoint, ordered_utxo.utxo.clone());
+                outpoint
+            })
+            .cloned()
+            .collect();
+
+        self.curr_buf.push_back((block.hash, block.height));
+        self.sent.insert(block.hash, outpoints);
+    }
+
+    /// Try to look up this UTXO in any sent block.
+    #[instrument(skip(self))]
+    pub fn utxo(&self, outpoint: &transparent::OutPoint) -> Option<transparent::Utxo> {
+        self.known_utxos.get(outpoint).cloned()
+    }
+
+    /// Replaces `curr_buf` with an empty VecDeque and pushes the previous `curr_buf` to `sent_bufs`
```

Comment:
Nitpick: same as above.

```suggestion
    /// Finishes the current block batch, and stores it for efficient pruning.
```